index.d.ts

```diff
@@ -0,0 +1,60 @@
+export interface Action {
+  type: string;
+}
+
+export type Reducer<S> = (state: S, action: Action) => S;
+
+export type Dispatch = (action: any) => any;
```

Comment:
That's fine for default dispatch (no middleware):
```ts
type DefaultDispatch = <A extends Action>(action: A) => A;
```
The problem here is that dispatch doesn't always return its input. Consider `redux-thunk`:
```ts
store.dispatch(dispatch => 42)  // returns 42
```
After applying thunk middleware `dispatch` gets type:
```ts
type ThunkDispatch = <O>(thunk: (dispatch, getState) => O) => O;
typeof store.dispatch == DefaultDispatch | ThunkDispatch;
```
Things get worse when we try to define types for `thunk` arguments. In case of single `thunk` middleware it is obviously `DefaultDispatch`:
```ts
type ThunkDispatch =
  <O, S>(thunk: (dispatch: DefaultDispatch, getState: () => S) => O) => O;
```
But what if we had e.g. Promise middleware in front of `thunk` middleware? We should be able to dispatch promises from `thunk`:
```ts
type PromiseDispatch = <P extends Promise<any>>(promise: P): P;
type ThunkDispatch =
  <O, S>(thunk: (dispatch: DefaultDispatch | PromiseDispatch, getState: () => S) => O) => O;
```
This means that `thunk` middleware should know about previously applied middlewares.